### ISSUE TITLE
Invalid json after deleting collapsed tags

### ISSUE THREAD

**Issue by cdalexndr** [notepad-plus-plus/notepad-plus-plus#6127]
Consider the following json:
```
{
 "a":"b"
},
{
 "a":"b"
},
{
 "a":"b"
}
```
Set language to Json, and collapse all tags.
Select last two lines and press delete. Result:
```
{
 "a":"b"
},

 "a":"b"
}
```
Expected:
```
{
 "a":"b"
},
```
Notepad++ v7.7.1

**Comment by Ekopalypse** [comment]
Your expectation is what I get as the result.
Maybe you wanna explain in more detail the steps you do to get this behavior
and post your debug info.

**Comment by cdalexndr** [comment]
After more playing around, it seems there is a selection issue.

When tags are collapsed, if I click **on** the last line ending and drag above to add the above line to the selection:
![notepad_sel](https://user-images.githubusercontent.com/22418250/64345855-bc08a400-cff9-11e9-947b-da2118160321.png)
then deleting only removes starting tag.

If I click **below** the last line and drag to select both lines:
![notepad_sel2](https://user-images.githubusercontent.com/22418250/64345925-dd699000-cff9-11e9-86d0-8cebe5c77e3f.png)
then deleting works as expected.

This behavior is confusing, because in normal text editing when you drag a selection from a line ending and delete, the whole line will be deleted except the ending newline char. In this case, only the shown starting tag "{" is deleted and the rest hidden/collapsed content remains. I expected the same behavior in this case, to delete all collapsed json and keep a newline.

--------------------------------------
Notepad++ v7.7.1   (32-bit)
Build time : Jun 16 2019 - 21:14:50
Path : C:\Program Files (x86)\Notepad++\notepad++.exe
Admin mode : OFF
Local Conf mode : OFF
OS : Windows 10 (64-bit)
Plugins : DSpellCheck.dll mimeTools.dll NppConverter.dll NppExport.dll

**Comment by cdalexndr** [comment]
Tested some other editors that use the behavior I expected: intellij IDEA and https://jsongrid.com/json-viewer

**Comment by sasumner** [comment]
@cdalexndr If I try it in the Scintilla demo editor, `Sc1`, I _think_ I can duplicate your behavior.  You may want to try that editor as an additional test to confirm, as I am not quite sure from your description that I am doing exactly what you are doing.  Given that I'm fairly certain that I can duplicate it in Scintilla, I'm going to give this issue ticket the "Scintilla Dependent" label, meaning that it is something that can't be fixed in Notepad++ unless it is first addressed in Scintilla.

**Comment by cdalexndr** [comment]
Indeed, same unexpected behaviour in [SciTE Version 4.2.0  Jul  5 2019 08:28:50](https://www.scintilla.org/wscite420.zip).
Only "{" is deleted.

**Comment by ArkadiuszMichalski** [comment]
Hmm I check this closer and everything depend how we select this folded line. In https://github.com/notepad-plus-plus/notepad-plus-plus/issues/6127 we see that wilder (gray) range means correct behaviour (if we get such a selection, the lines delete properly). It all depends on whether we also select EOL when try selecting 2 (or more) collapsed lines.

If we start selecting from 4 to 7 line we must also select EOL (which means we have to extend the selection to the line below).
![c](https://user-images.githubusercontent.com/2730894/94088839-c159ac00-fe11-11ea-8d21-ecf96060d4c9.gif)

But the same is when we start selecting from 7-4 (for selecting EOL we must start belowe line 7) 
![d](https://user-images.githubusercontent.com/2730894/94088963-17c6ea80-fe12-11ea-8c10-f229f71c0bff.gif)

For uncolapsed all we have:
![c](https://user-images.githubusercontent.com/2730894/94089325-2feb3980-fe13-11ea-9b4f-f3b89c7be844.gif)

This is weird, we get diffrent result depending on how we make selection/deletion in collapsed line/lines:
- select and delete only `{` in single collapsed line
- select and delete only EOL in single collapsed line
- select `{` and or without EOL in multi collapsed line << this case looks bugy or maybe it's future:)

**Comment by ArkadiuszMichalski** [comment]
I check this also for HTML and the same, looks like this isn't limited only to JSON. 
```
<p>
A
</p>
<div>
B
</div>
<ul>
C
</ul>
```
![a](https://user-images.githubusercontent.com/2730894/94117630-c1c36880-fe4c-11ea-9395-ef889d46ada2.gif)

I will not describe all the situation when e.g. we make a few selections with Ctrl, here are also some nice cases, try:
- select only P+DIV+UL (without any EOL) and remove
- select only P+UL  (without any EOL) and remove
- select only P (with EOL) + UL (without EOL) and remove
- select only DIV (with EOL) + UL (without EOL) and remove
...

**Comment by vinsworldcom** [comment]
@donho
@sasumner

I can confirm this is NOT fixed in Notepad++ 7.9.3 Unicode Debug using the latest Scintilla 4.4.6 DLL.

I remember addressing whether the line the cursor was on (if it was at the beginning) should or should not be involved in the selection with [this commit](https://github.com/notepad-plus-plus/notepad-plus-plus/commit/7c28a120d953d03a514100609bee2c4ab1b63021) for indenting lines.  Could the same be applied here for deleting lines perhaps?

Cheers.
